compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -3431,9 +3430,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 })
                                 // For compile-time reasons put a limit on number of fields we search
                                 .take(100)
+                                .map(|field_def| {
+                                    (
+                                        field_def.ident(self.tcx).normalize_to_macros_2_0(),
+                                        field_def.ty(self.tcx, args),
+                                    )
+                                })
+                                .collect::<Vec<_>>(),
+                        );
+                    }
+                    ty::Tuple(types) => {
+                        return Some(
+                            types
+                                .iter()
+                                .enumerate()
+                                // For compile-time reasons put a limit on number of fields we search
+                                .take(100)
```

Comment:
can you think of any (even-non practical) test where this matters? If there are enough fields for a linear search to cause performance issues, I'd assume that we've already got performance issues due to other reasons, e.g. trying to print the full type name to a file

Reply:
1) the check was already present already
2) I think it is good form to make diagnostics not incur additional time during compiles, within reason
3) Ithis logic could trigger multiple times, up to 3 levels deep: if you have 3 levels of nesting where all of the types have 100 fields, this is already going to be slow (1M ops); making it unbounded could end up with the compiler hanging in a case where it otherwise wouldn't